Rework readme (#86)
Sensirion/embedded-common: i2c/README.md

# Getting started

## Implement the I2C Interface

So we need to adjust two files according to your platform.

### Edit `sensirion_i2c_hal.c`

> This file contains the implementation of the sensor communication (how to send
> requests to the sensor). Therefore, how this is done depends on your hardware
> platform. Therefore we can only provide function stubs in which you can
> implement the logic yourself.
> There are sample implementations available for some platforms:
> [`sample-implementations`](sample-implementations). If you are using a Linux
> based platform like Raspberry Pi you can just replace the unimplemented HAL
> template with the provided implementation in
> `sample-implementations/linux_user_space/`:

Reviewer: ```suggestion
This file contains the implementation of the sensor communication, which
depends on your hardware platform. We provide function stubs for your
hardware's own implementation.
Sample implementations are available for some platforms:
[`sample-implementations`](sample-implementations). For Linux
based platforms like Raspberry Pi you can just replace the unimplemented HAL
template with the implementation in
`sample-implementations/linux_user_space/`:
```

```
cp sample-implementations/linux_user_space/sensirion_i2c_hal.c ./
```

### Edit `sensirion_config.h`

> If you are on a Linux based platform you can skip this part since everything is
> already correctly setup for you.

Reviewer: ```suggestion
Skip this part for Linux based platforms since everything is already setup for this case.
```

Otherwise you need to check if the libraries `<stdint.h>` and `<stdlib.h>` are
provided by your toolchain, compiler or system. If you have no idea on how to
do that you can skip this step for now and come back when you get errors
related to these names when compiling the driver.

> The features we use out of those libraries are standard integer sizes from

Reviewer: ```suggestion
The features we use from those libraries are type definitions for integer sizes from
```

* `int8_t` = signed 8bit integer
* `uint8_t` = unsigned 8bit integer

In addition to that you will need to specify `NULL`. For both we have a
detailed template where you just need to fill in your system specific values.

Now we are ready to compile and run the example usage for your sensor.

## Compile and Run

> Take the `.c` and `.h` files directly in this folder pass them to your favorite

Reviewer: ```suggestion
Pass the source `.c` and header `.h` files in this folder into your
```

C compiler and run the resulting binary. This step may vary, depending on your
platform. Here we demonstrate the procedure for Linux based platforms:

1. Open up a terminal.
2. Navigate to the directory where this README is located.

> 3. Run `make` (this compiles all the code here to one binary).

Reviewer: ```suggestion
3. Run `make` (this compiles the example code into one executable binary).
```

> 4. Run `./[SENSORNAME]_i2c_example_usage` (This will run your newly compiled
>    binary).

Reviewer: ```suggestion
4. Run the compiled executable with `./[SENSORNAME]_i2c_example_usage`
```

5. Now you should see the first measurement values appear in your terminal. As
   a next step you can adjust the example usage file or write your own main
   function to use the sensor.

# Background

## Files

### sensirion\_i2c.[ch]

In these files you can find the implementation of the I2C protocol used by
Sensirion sensors. The functions in these files are used by the embedded driver
to build the correct frame out of data to be sent to the sensor or receive a
frame of data from the sensor and convert it back to data readable by your
machine. The functions in here calculate and check CRCs, reorder bytes for
different byte orders and build the correct formatted frame for your sensor.

### sensirion\_i2c\_hal.[ch]

> In these files you can find the implementation of the hardware abstraction

Reviewer: ```suggestion
These files contain the implementation of the hardware abstraction
```
